internal/provider/resource_custom_location.go

```diff
@@ -0,0 +1,219 @@
+package cpln
+
+import (
+	"context"
+
+	client "github.com/controlplane-com/terraform-provider-cpln/internal/provider/client"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func resourceCustomLocation() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceCustomLocationCreate,
+		ReadContext:   resourceCustomLocationRead,
+		UpdateContext: resourceCustomLocationUpdate,
+		DeleteContext: resourceCustomLocationDelete,
+		Schema: map[string]*schema.Schema{
+			"cpln_id": {
+				Type:        schema.TypeString,
+				Description: "The ID, in GUID format, of the location.",
+				Computed:    true,
+			},
+			"name": {
+				Type:         schema.TypeString,
+				Description:  "Name of the Location.",
+				ForceNew:     true,
+				Required:     true,
+				ValidateFunc: NameValidator,
+			},
+			"description": {
+				Type:             schema.TypeString,
+				Description:      "Description of the location.",
+				Optional:         true,
+				ValidateFunc:     DescriptionValidator,
+				DiffSuppressFunc: DiffSuppressDescription,
+			},
+			"tags": {
+				Type:        schema.TypeMap,
+				Description: "Key-value map of resource tags.",
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Optional:     true,
+				ValidateFunc: TagValidator,
+			},
+			"cloud_provider": {
+				Type:        schema.TypeString,
+				Description: "Cloud Provider of the location.",
+				Required:    true,
+			},
+			"enabled": {
+				Type:        schema.TypeBool,
+				Description: "Indication if location is enabled.",
+				Required:    true,
+			},
+			"ip_ranges": {
+				Type:        schema.TypeSet,
+				Description: "A list of IP ranges of the location.",
+				Computed:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"self_link": {
+				Type:        schema.TypeString,
+				Description: "Full link to this resource. Can be referenced by other resources.",
+				Computed:    true,
+			},
+		},
+		Importer: &schema.ResourceImporter{},
+	}
+}
+
+func resourceCustomLocationCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+
+	c := m.(*client.Client)
+
+	// Define & Build
+	locationName := d.Get("name").(string)
+	isLocationEnabled := d.Get("enabled").(bool)
+	description := GetString(d.Get("description").(string))
+	tags := GetStringMap(d.Get("tags"))
+	provider := d.Get("cloud_provider").(string)
+
+	if provider != "byok" {
+		return diag.Errorf("provider must be byok for custom locations")
```

Comment:
Change the error message to: `cloud_provider must be 'byok' for custom locations`